### PR TITLE
fix(twin,amqp): refactor AMQP implementation of the twin

### PR DESCRIPTION
**Pull request by pierreca**
This PR contains refactoring for the AMQP implementation of the twin.
The previous one contained the MQTT implementation and the next one will take care of the client.
All will be merged together as one.

### CODE REVIEW

**Comment by BertKleewein** [comment]
>   "name": "azure-iot-device-amqp",

Test to verify that I am able to comment. #Closed

---
Refers to: device/transport/amqp/package.json:2 in b86caab. [](commit_id = b86caab79493072eef3e700232d0f6bce0dfff5b, deletion_comment = False)

File: device/transport/amqp/devdoc/amqp_twin_client_requirements.md [hunk 1]
```diff
@@ -3,106 +3,139 @@
...
-### Constructor
+### getTwin(callback: (err: Error, twin?: TwinProperties) => void): void;
 
-**SRS_NODE_DEVICE_AMQP_TWIN_06_004: [** The `AmqpTwinClient` constructor shall throw `ReferenceError` if the `config` object is falsy. **]**
+**SRS_NODE_DEVICE_AMQP_TWIN_16_007: [** The `getTwin` method shall attach the sender link if it's not already attached. **]**
 
-**SRS_NODE_DEVICE_AMQP_TWIN_06_002: [** The `AmqpTwinClient` constructor shall throw `ReferenceError` if the `client` object is falsy. **]**
+**SRS_NODE_DEVICE_AMQP_TWIN_16_008: [** If attaching the sender link fails, the `getTwin` method shall call its callback with the error that caused the failure. **]**
 
-**SRS_NODE_DEVICE_AMQP_TWIN_06_005: [** The `AmqpDeviceMethodClient` shall inherit from the `EventEmitter` class. **]**
```
Comment (anthonyvercolano) [inline device/transport/amqp/devdoc/amqp_twin_client_requirements.md:24]: >AmqpDeviceMethodClien [](start = 45, length = 21)

Testing that I can post a comment.  I can!!  Although one could ask, how did methodClient get through here! #Closed

File: device/transport/amqp/devdoc/amqp_twin_client_requirements.md [hunk 2]
```diff
@@ -3,106 +3,139 @@
...
-var receiver = new AmqpTwinClient(config: ClientConfig, client: any);
-receiver.on('response', function(response) {
-    console.log('Response received for request ' + response.requestId);
-    console.log('  status = " + response.status);
-    console.log(response.body);
-});
+## Public API
+
+```typescript
+class AmqpTwinClient extends EventEmitter {
```
Comment (BertKleewein) [inline device/transport/amqp/devdoc/amqp_twin_client_requirements.md:9]: >AmqpTwinClient [](start = 6, length = 14)

I haven't gone back to the MQTT, but is this the same interface that MQTT uses?  If so, do we want to codify it as an interface? #ByDesign

File: device/transport/amqp/devdoc/amqp_twin_client_requirements.md [hunk 3]
```diff
@@ -3,106 +3,139 @@
...
+  updateTwinReportedProperties(patch: any, callback: (err?: Error) => void): void;
+  enableTwinDesiredPropertiesUpdates(callback: (err?: Error) => void): void;
+  disableTwinDesiredPropertiesUpdates(callback: (err?: Error) => void): void;
+  detach(callback: (err?: Error) => void): void;
+}
 ```
+### constructor(authenticationProvider: AuthenticationProvider, client: any) {
 
-## Public API
+**SRS_NODE_DEVICE_AMQP_TWIN_06_005: [** The `AmqpDeviceMethodClient` shall inherit from the `EventEmitter` class. **]**
```
Comment (BertKleewein) [inline device/transport/amqp/devdoc/amqp_twin_client_requirements.md]: >SRS_NODE_DEVICE_AMQP_TWIN_06_005 [](start = 2, length = 32)

OK, but this seems out of scope for what we usually specify.  This seems like an implementation detail or a comment about how the code is structured, not something about how the code behaves.
 #ByDesign

Comment (BertKleewein) [inline device/transport/amqp/devdoc/amqp_twin_client_requirements.md]: >AmqpDeviceMethodClient [](start = 45, length = 22)

AmqpTwinClient #Resolved

File: device/transport/amqp/devdoc/amqp_twin_client_requirements.md [hunk 4]
```diff
@@ -3,106 +3,139 @@
...
+**SRS_NODE_DEVICE_AMQP_TWIN_16_009: [** THe `getTwin` method shall attach the receiver link if it's not already attached. **]**
 
-### response event
+**SRS_NODE_DEVICE_AMQP_TWIN_16_010: [** If attaching the receiver link fails, the `getTwin` method shall call its callback with the error that caused the failure. **]**
 
-**SRS_NODE_DEVICE_AMQP_TWIN_06_006: [** When a listener is added for the `response` event, and the `post` event is NOT already subscribed, upstream and downstream links are established via calls to `attachReceiverLink` and `attachSenderLink`. **]**
+**SRS_NODE_DEVICE_AMQP_TWIN_16_011: [** The `getTwin` method shall send an `AmqpMessage` using the `SenderLink.send` method with the following annotations and properties:
+- `operation` annotation set to `GET`.
+- `resource` annotation set to `undefined`
+- `correlationId` property set to a random integer
```
Comment (BertKleewein) [inline device/transport/amqp/devdoc/amqp_twin_client_requirements.md]: >integer [](start = 43, length = 7)

guid? #Resolved

File: device/transport/amqp/devdoc/amqp_twin_client_requirements.md [hunk 5]
```diff
@@ -3,106 +3,139 @@
...
+- `body` set to ` `. **]**
 
-**SRS_NODE_DEVICE_AMQP_TWIN_06_007: [** The endpoint argument for attacheReceiverLink shall be `/devices/<deviceId>/twin`. **]**
+**SRS_NODE_DEVICE_AMQP_TWIN_16_012: [** If the `SenderLink.send` call fails the `getTwin` method shall call its callback with the error that caused the failure. **]**
```
Comment (BertKleewein) [inline device/transport/amqp/devdoc/amqp_twin_client_requirements.md]: >If the `SenderLink.send` call fails the `getTwin` method [](start = 40, length = 56)

nit: add comma after the word "fails".  Otherwise, it first reads as the getTwin method being the object of the failure, not the subject of the following clause. #Resolved

File: device/transport/amqp/devdoc/amqp_twin_client_requirements.md [hunk 6]
```diff
@@ -3,106 +3,139 @@
...
-**SRS_NODE_DEVICE_AMQP_TWIN_06_008: [** The link options argument for attachReceiverLink shall be:
- attach: {
-        properties: {
-          'com.microsoft:channel-correlation-id' : 'twin:<correlationId>',
-          'com.microsoft:api-version' : endpoint.apiVersion
-        },
-        sndSettleMode: 1,
-        rcvSettleMode: 0
-      } **]**
+**SRS_NODE_DEVICE_AMQP_TWIN_16_013: [** The `getTwin` method shall monitor `Message` objects on the `ReceiverLink.on('message')` handler and until a message with the same `correlationId` as the one that was sent is received. **]**
```
Comment (BertKleewein) [inline device/transport/amqp/devdoc/amqp_twin_client_requirements.md]: > handler and until a message [](start = 128, length = 28)

nit: this doesn't parse.  I _think_ the word "and" was accidently left in. #Resolved

File: device/transport/amqp/devdoc/amqp_twin_client_requirements.md [hunk 7]
```diff
@@ -3,106 +3,139 @@
...
 
+**SRS_NODE_DEVICE_AMQP_TWIN_16_014: [** The `getTwin` method shall parse the body of the received message and call its callback with a `null` error object and the parsed object as a result. **]**
```
Comment (BertKleewein) [inline device/transport/amqp/devdoc/amqp_twin_client_requirements.md]: >call its callback with a `null` error object  [](start = 110, length = 45)

I think this is what you were referring to when you mentioned twin code not checking the status code, but, if not, it looks like this code doesn't check the status code :) #WontFix

File: device/transport/amqp/devdoc/amqp_twin_client_requirements.md [hunk 8]
```diff
@@ -3,106 +3,139 @@
...
+- `operation` annotation set to `PATCH`.
+- `resource` annotation set to `/properties/reported`
+- `correlationId` property set to a random integer
+- `body` set to the stringified patch object. **]**
 
-**SRS_NODE_DEVICE_AMQP_TWIN_06_017: [** The `requestId` value is acquired from the amqp message correlationId property in the response amqp message. **]**
+**SRS_NODE_DEVICE_AMQP_TWIN_16_020: [** If the `SenderLink.send` call fails the `updateTwinReportedProperties` method shall call its callback with the error that caused the failure. **]**
 
-**SRS_NODE_DEVICE_AMQP_TWIN_06_026: [** The `status` value is acquired from the amqp message status message annotation. **]**
+**SRS_NODE_DEVICE_AMQP_TWIN_16_021: [** The `updateTwinReportedProperties` method shall monitor `Message` objects on the `ReceiverLink.on('message')` handler and until a message with the same `correlationId` as the one that was sent is received. **]**
```
Comment (BertKleewein) [inline device/transport/amqp/devdoc/amqp_twin_client_requirements.md]: >handler and until [](start = 150, length = 17)

same parsing error as 16_013 #Resolved

File: device/transport/amqp/devdoc/amqp_twin_client_requirements.md [hunk 9]
```diff
@@ -3,106 +3,139 @@
...
+- `operation` annotation set to `PUT`.
+- `resource` annotation set to `/notifications/twin/properties/desired`
+- `correlationId` property set to a random integer
+- `body` set to `undefined`. **]**
 
-### error
+**SRS_NODE_DEVICE_AMQP_TWIN_16_028: [** If the `SenderLink.send` call fails the `enableTwinDesiredPropertiesUpdates` method shall call its callback with the error that caused the failure. **]**
 
-**SRS_NODE_DEVICE_AMQP_TWIN_06_022: [** If an error occurs on establishing the upstream or downstream link then the `error` event shall be emitted. **]**
+**SRS_NODE_DEVICE_AMQP_TWIN_16_029: [** The `enableTwinDesiredPropertiesUpdates` method shall monitor `Message` objects on the `ReceiverLink.on('message')` handler and until a message with the same `correlationId` as the one that was sent is received. **]**
```
Comment (BertKleewein) [inline device/transport/amqp/devdoc/amqp_twin_client_requirements.md]: >handler and until  [](start = 156, length = 18)

same parsing error again #Resolved

File: device/transport/amqp/devdoc/amqp_twin_client_requirements.md [hunk 10]
```diff
@@ -3,106 +3,139 @@
...
 
-**SRS_NODE_DEVICE_AMQP_TWIN_16_007: [** The `attach` method shall call the `getDeviceCredentials` method on the `authenticationProvider` object passed as an argument to the constructor to retrieve the device id. **]**
+**SRS_NODE_DEVICE_AMQP_TWIN_16_034: [** The `disableTwinDesiredPropertiesUpdates` method shall monitor `Message` objects on the `ReceiverLink.on('message')` handler and until a message with the same `correlationId` as the one that was sent is received. **]**
 
-**SRS_NODE_DEVICE_AMQP_TWIN_16_008: [** The `attach` method shall call its callback with an error if the call to `getDeviceCredentials` fails with an error. **]**
+**SRS_NODE_DEVICE_AMQP_TWIN_16_035: [** The `disableTwinDesiredPropertiesUpdates` method shall call its callback with no argument when a response is received **]**
+// TODO: we need to check error codes!
 
-### detach(callback)
+### detach(callback: (err?: Error) => void): void;
```
Comment (BertKleewein) [inline device/transport/amqp/devdoc/amqp_twin_client_requirements.md:108]: >detach [](start = 4, length = 6)

possibly silly question: should a clean detach do a DELETE operation on /notifications/twin/properties/desired like we're being told to do for MQTT? #Closed

File: device/transport/amqp/devdoc/amqp_twin_client_requirements.md [hunk 11]
```diff
@@ -3,106 +3,139 @@
...
 
 **SRS_NODE_DEVICE_AMQP_TWIN_16_004: [** The `detach` method shall call its `callback` immediately if the links are already detached. **]**
 
 **SRS_NODE_DEVICE_AMQP_TWIN_16_005: [** The `detach` method shall detach the links and call its `callback` with no arguments if the links are successfully detached. **]**
 
 **SRS_NODE_DEVICE_AMQP_TWIN_16_006: [** The `detach` method shall call its `callback` with an `Error` if detaching either of the links fail. **]**
+
+### Links
+
+**SRS_NODE_DEVICE_AMQP_TWIN_06_007: [** The endpoint argument for attacheReceiverLink shall be `/devices/<deviceId>/twin`. **]**
```
Comment (BertKleewein) [inline device/transport/amqp/devdoc/amqp_twin_client_requirements.md]: >attacheReceiverLink [](start = 66, length = 19)

typo: s/ attache/attach/ #Resolved

File: device/transport/amqp/devdoc/amqp_twin_client_requirements.md [hunk 12]
```diff
@@ -3,106 +3,139 @@
...
+        rcvSettleMode: 0
+      } **]**
+
+
+**SRS_NODE_DEVICE_AMQP_TWIN_06_009: [** The endpoint argument for attacheSenderLink shall be `/device/<deviceId>/twin`. **]**
+
+**SRS_NODE_DEVICE_AMQP_TWIN_06_010: [** The link options argument for attachSenderLink shall be:
+ attach: {
+        properties: {
+          'com.microsoft:channel-correlation-id' : 'twin:<correlationId>',
```
Comment (BertKleewein) [inline device/transport/amqp/devdoc/amqp_twin_client_requirements.md]: >twin:<correlationId> [](start = 52, length = 20)

I assume the correlationId needs to be the same between the senderLink and the receiverLink, right?  If so, do we need to specify this requirement?  #Resolved

**Comment by BertKleewein** [comment]
> const handleResult = function (errorMessage: string, done: (err?: Error, result?: any) => void): (err?: Error, result?: any) => void {

oooh, this is nifty.  I'll have to remember this. #Resolved

---
Refers to: device/transport/amqp/src/amqp.ts:21 in b86caab. [](commit_id = b86caab79493072eef3e700232d0f6bce0dfff5b, deletion_comment = False)

File: device/transport/amqp/src/amqp_twin_client.ts [hunk 13]
```diff
@@ -6,102 +6,134 @@
...
   private _endpoint: string;
-  private _upstreamAmqpLink: any;
-  private _downstreamAmqpLink: any;
+  private _senderLink: SenderLink;
+  private _receiverLink: ReceiverLink;
   private _fsm: any;
-  private _internalOperations: { [key: string]: () => void };
-  private _eventQueue: any[];
-  private _eventQueueError: Error;
+  // need type for twin properties
```
Comment (BertKleewein) [inline device/transport/amqp/src/amqp_twin_client.ts]: Mark with TODO: if this is a todo.  Otherwise, it looks like you're trying to describe what the code does #Resolved

File: device/transport/amqp/src/amqp_twin_client.ts [hunk 14]
```diff
@@ -6,102 +6,134 @@
...
+  private _pendingTwinRequests: { [key: string]: (err: Error, twinProperties?: any) => void };
+
+  private _messageHandler: (message: Message) => void;
+  private _errorHandler: (err: Error) => void;
 
   constructor(authenticationProvider: AuthenticationProvider, client: any) {
```
Comment (BertKleewein) [inline device/transport/amqp/src/amqp_twin_client.ts:49]: >authenticationProvider: AuthenticationProvider [](start = 14, length = 46)

out of scope, but I'm wondering why you don't just pass the credentials down?  Why pass a higher-level construct when a lower-level one will do just as well? #WontFix

File: device/transport/amqp/src/amqp_twin_client.ts [hunk 15]
```diff
@@ -118,16 +150,17 @@ export class AmqpTwinClient extends EventEmitter {
...
                     } else {
-                      this._downstreamAmqpLink = receiverTransportObject;
-                      this._downstreamAmqpLink.on('error', this._handleError.bind(this));
+                      this._receiverLink = receiverTransportObject;
+                      this._receiverLink.on('message', this._messageHandler);
+                      this._receiverLink.on('error', this._errorHandler);
                       this._client.attachSenderLink( this._endpoint, this._generateTwinLinkProperties(linkCorrelationId), (senderLinkError?: Error, senderTransportObject?: any): void => {
                         if (senderLinkError) {
-                          this._fsm.handle('handleErrorEmit', senderLinkError);
+                          this._fsm.transition('detached', senderLinkError, attachCallback);
```
Comment (BertKleewein) [inline device/transport/amqp/src/amqp_twin_client.ts:159]: >transition [](start = 36, length = 10)

detach the receiver link if the sender link fails to attach? #WontFix

File: device/transport/amqp/src/amqp_twin_client.ts [hunk 16]
```diff
@@ -136,164 +169,96 @@ export class AmqpTwinClient extends EventEmitter {
...
+  /**
+   * Necessary for the client to be able to properly detach twin links
+   * attach() isn't necessary because it's done by the FSM automatically when one of the APIs is called.
+   */
+  detach(callback: (err?: Error) => void): void {
     this._fsm.handle('detach', callback);
   }
 
+
   private _generateTwinLinkProperties( correlationId: string): any {
```
Comment (BertKleewein) [inline device/transport/amqp/src/amqp_twin_client.ts]: >_generateTwinLinkProperties [](start = 10, length = 27)

Do my eyes deceive me, or is there whitespace above this line? #Resolved

File: device/transport/amqp/test/_amqp_twin_client_test.js [hunk 17]
```diff
@@ -93,687 +113,448 @@ describe('AmqpTwinClient', function () {
...
-            newMessage.data = messageData;
-            newMessage.transportObj = {};
-            newMessage.transportObj.messageAnnotations = {};
-            newMessage.transportObj.messageAnnotations['status'] = 201;
-            amqpClient.fakeReceiverLink.emit('message', newMessage);
-          });
-        }
+    it('configures link options for the sender and receiver links', function() {
+      methodUnderTest();
+      assert.equal(fakeAmqpClient.attachSenderLink.firstCall.args[1].attach.sndSettleMode, 1, ' sender send settle mode not set appropriately' );
```
Comment (BertKleewein) [inline device/transport/amqp/test/_amqp_twin_client_test.js]: >fakeAmqpClient.attachSenderLink.firstCall.args[1].attach [](start = 19, length = 56)

nit: Looks like a perfect place to use a temporary to hold fakeAmqpClient.attachSenderLink.firstCall.args[1].attach and fakeAmqpClient.attachReceiver.firstCall.args[1].attach :) #Resolved

File: device/transport/amqp/test/_amqp_twin_client_test.js [hunk 18]
```diff
@@ -93,687 +113,448 @@ describe('AmqpTwinClient', function () {
...
-        newTwinReceiver.on('error', function(err) {
-          assert.equal(err.constructor.name, 'UnauthorizedError');
-          testDone();
-        });
-        newTwinReceiver.on('response', function () {});
-      });
+    it('sends a message with an empty body', function () {
+      twinClient.getTwin(function () {});
+      assert.strictEqual(fakeSenderLink.send.firstCall.args[0].body, ' ');
     });
```
Comment (BertKleewein) [inline device/transport/amqp/test/_amqp_twin_client_test.js]: Do you get paid by the test?  These 4 could be 1.  #Resolved

Comment (BertKleewein) [reply device/transport/amqp/test/_amqp_twin_client_test.js]: add nit:

---
In reply to: [171690404](https://github.com/pierreca/azure-iot-sdk-node/pull/1#discussion_r171690404) [](ancestors = 171690404)

File: device/transport/amqp/test/_amqp_twin_client_test.js [hunk 19]
```diff
@@ -93,687 +113,448 @@ describe('AmqpTwinClient', function () {
...
-      var newTwinReceiver = new AmqpTwinClient(fakeAuthenticationProvider, amqpClient);
-      newTwinReceiver.sendTwinRequest('PATCH', '/properties/reported/', { '$rid':10, '$version': 200 }, 'body', function(err, res) {
-        if (err) {
-          done(err);
-        } else {
-          assert.equal(res.constructor.name, 'MessageEnqueued');
-          done();
-        }
-      });
+    it('sends a message with the body set to the stringified patch content', function () {
```
Comment (BertKleewein) [inline device/transport/amqp/test/_amqp_twin_client_test.js]: >it [](start = 4, length = 2)

nit 4 tests could be 1 #Resolved

File: device/transport/amqp/test/_amqp_twin_client_test.js [hunk 20]
```diff
@@ -93,687 +113,448 @@ describe('AmqpTwinClient', function () {
...
+    });
+
+    it('sends a message with a correlation id that is an integer', function () {
+      twinClient.enableTwinDesiredPropertiesUpdates(function () {});
+      var correlationId = parseInt(fakeSenderLink.send.firstCall.args[0].properties.correlationId);
+      assert.isNumber(correlationId);
+      assert.isNotNaN(correlationId);
+    });
+
+    it('sends a message with an empty body', function () {
```
Comment (BertKleewein) [inline device/transport/amqp/test/_amqp_twin_client_test.js]: >(' [](start = 6, length = 2)

nit: 4 could be 1 #Resolved

File: device/transport/amqp/devdoc/amqp_twin_client_requirements.md (hunk 10 above)
Comment (BertKleewein) [reply device/transport/amqp/devdoc/amqp_twin_client_requirements.md:108]: OK, it looks like the code does this, and I see 16_005 now.   Oops.

---
In reply to: [171684486](https://github.com/pierreca/azure-iot-sdk-node/pull/1#discussion_r171684486) [](ancestors = 171684486)

**Review by BertKleewein — APPROVED** [review]
:shipit:

File: device/transport/amqp/devdoc/amqp_twin_client_requirements.md (hunk 2 above)
Comment (pierreca) [reply device/transport/amqp/devdoc/amqp_twin_client_requirements.md:9]: it sort of is but since these are "hidden" in the transport i'm not sure where i'd put this or what purpose it would serve. each transport might evolve independently and as you said since there is a smell associated with a separate twin client if we end up changing the pattern i would rather have decoupled transports,

---
In reply to: [171681233](https://github.com/pierreca/azure-iot-sdk-node/pull/1#discussion_r171681233) [](ancestors = 171681233)

File: device/transport/amqp/devdoc/amqp_twin_client_requirements.md (hunk 3 above)
Comment (pierreca) [reply device/transport/amqp/devdoc/amqp_twin_client_requirements.md]: yes and no, it's also a good way to carry the idea that this thing is going to fire events I think we have it in other places in the code too for other event emitters.

---
In reply to: [171681955](https://github.com/pierreca/azure-iot-sdk-node/pull/1#discussion_r171681955) [](ancestors = 171681955)

File: device/transport/amqp/devdoc/amqp_twin_client_requirements.md (hunk 7 above)
Comment (pierreca) [reply device/transport/amqp/devdoc/amqp_twin_client_requirements.md]: yeah and since the annotations are only surfaced through the transportObj i'm not sure how to do this cleanly...

---
In reply to: [171683343](https://github.com/pierreca/azure-iot-sdk-node/pull/1#discussion_r171683343) [](ancestors = 171683343)

File: device/transport/amqp/devdoc/amqp_twin_client_requirements.md (hunk 12 above)
Comment (pierreca) [reply device/transport/amqp/devdoc/amqp_twin_client_requirements.md]: we do test for this. but no requirement. I'll add one.

---
In reply to: [171685162](https://github.com/pierreca/azure-iot-sdk-node/pull/1#discussion_r171685162) [](ancestors = 171685162)

File: device/transport/amqp/src/amqp_twin_client.ts (hunk 14 above)
Comment (pierreca) [reply device/transport/amqp/src/amqp_twin_client.ts:49]: Yeah it looks like we could just pass down the device id. i'll bug that.

---
In reply to: [171687202](https://github.com/pierreca/azure-iot-sdk-node/pull/1#discussion_r171687202) [](ancestors = 171687202)

File: device/transport/amqp/src/amqp_twin_client.ts (hunk 15 above)
Comment (pierreca) [reply device/transport/amqp/src/amqp_twin_client.ts:159]: yes in theory. in practice if a link is being detached the whole thing is going down and detach is going to be handled by the lower layer. I'm not satisfied with how we carry links around in the AMQP stack for now but this would be the subject of a bigger refactoring. 

---
In reply to: [171687889](https://github.com/pierreca/azure-iot-sdk-node/pull/1#discussion_r171687889) [](ancestors = 171687889)

File: device/transport/amqp/src/amqp_twin_client.ts (hunk 16 above)
Comment (pierreca) [reply device/transport/amqp/src/amqp_twin_client.ts]: damnit!

---
In reply to: [171688487](https://github.com/pierreca/azure-iot-sdk-node/pull/1#discussion_r171688487) [](ancestors = 171688487)

File: device/transport/amqp/test/_amqp_twin_client_test.js (hunk 18 above)
Comment (pierreca) [reply device/transport/amqp/test/_amqp_twin_client_test.js]: i always wonder about this: we're testing 4 different things technically. one could break independently of the others. but at the same time having 4 tests instead of 1 doesn't bring a lot of value. I guess I'll just make this one bigger test

---
In reply to: [171690468](https://github.com/pierreca/azure-iot-sdk-node/pull/1#discussion_r171690468) [](ancestors = 171690468,171690404)

File: device/transport/amqp/devdoc/amqp_twin_client_requirements.md (hunk 7 above)
Comment (pierreca) [reply device/transport/amqp/devdoc/amqp_twin_client_requirements.md]: I've bugged it and will address this separately

---
In reply to: [171741000](https://github.com/pierreca/azure-iot-sdk-node/pull/1#discussion_r171741000) [](ancestors = 171741000,171683343)
